pageserver/src/tenant.rs

```diff
@@ -1421,29 +1511,90 @@ impl Tenant {
                 }
             }
 
-            // Ensure that there are no non-archived child timelines
-            let children: Vec<TimelineId> = timelines
-                .iter()
-                .filter_map(|(id, entry)| {
-                    if entry.get_ancestor_timeline_id() != Some(timeline_id) {
-                        return None;
-                    }
-                    if entry.is_archived() == Some(true) {
-                        return None;
-                    }
-                    Some(*id)
-                })
-                .collect();
+            if new_state == TimelineArchivalState::Archived {
+                // Ensure that there are no non-archived child timelines
+                let children: Vec<TimelineId> = timelines
+                    .iter()
+                    .filter_map(|(id, entry)| {
+                        if entry.get_ancestor_timeline_id() != Some(timeline_id) {
+                            return None;
+                        }
+                        if entry.is_archived() == Some(true) {
+                            return None;
+                        }
+                        Some(*id)
+                    })
+                    .collect();
+
+                if !children.is_empty() {
+                    return Err(TimelineArchivalError::HasUnarchivedChildren(children));
+                }
+            }
+            Some(Arc::clone(timeline))
+        };
+
+        let timeline = if let Some(timeline) = timeline_or_unarchive_offloaded {
+            timeline
+        } else {
+            let cancel = self.cancel.clone();
```

Comment:
Can we make it more clear that this branch does the unarchive? Wasn't obvious until I read the archival path.

Reply:
I've tried to indicate that by the `timeline_or_unarchive_offloaded` variable name. But I can add a comment.